File: 990 Diving for Gold/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct {
    int d, v;
} Treasure;

int main() {
	int t, w;
	bool firstCase = true;
	while(scanf("%d %d", &t, &w) == 2) {
	    if(!firstCase)
	        putchar('\n');
	    int n;
	    scanf("%d", &n);
	    Treasure treasures[n];
	    for(int i = 0; i < n; ++i)
	        scanf("%d %d", &treasures[i].d, &treasures[i].v);
	    int dp[t + 1];
	    memset(dp, 0, sizeof(dp));
	    bool selected[n][t + 1];
	    memset(selected, 0, sizeof(selected));
	    for(int i = 0; i < n; ++i) {
	        int d = 3 * w * treasures[i].d;
	        int v = treasures[i].v;
	        for(int j = t; j >= d; --j) {
	            if(dp[j] < dp[j - d] + v) {
	                dp[j] = dp[j - d] + v;
	                selected[i][j] = true;
	            }
	        }
	    }
	    printf("%d\n", dp[t]);
	    int currentT = t, *treasureID = NULL, size = 0;
	    for(int i = n - 1; i >= 0; --i) {
	        if(selected[i][currentT]) {
	            treasureID = (int*)realloc(treasureID, (size + 1) * sizeof(int));
	            treasureID[size++] = i;
	            currentT -= 3 * w * treasures[i].d;
	        }
	    }
	    printf("%d\n", size);
	    for(int i = size - 1; i >= 0; --i)
	        printf("%d %d\n", treasures[treasureID[i]].d, treasures[treasureID[i]].v);
	    free(treasureID);
	    firstCase = false;
	}
	return 0;
}

File: 990 Diving for Gold/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
        StreamTokenizer st = new StreamTokenizer(System.in);
        boolean firstCase = true;
        StringBuilder output = new StringBuilder();
        while(st.nextToken() == StreamTokenizer.TT_NUMBER) {
            if(!firstCase)
                output.append('\n');
            int t = (int)st.nval;
            st.nextToken();
            int w = (int)st.nval;
            st.nextToken();
            int n = (int)st.nval;
            Treasure[] treasures = new Treasure[n];
            for(int i = 0; i < n; ++i) {
                treasures[i] = new Treasure();
                st.nextToken();
                treasures[i].d = (int)st.nval;
                st.nextToken();
                treasures[i].v = (int)st.nval;
            }
            int[] dp = new int[t + 1];
            boolean[][] selected = new boolean[n][t + 1];
            for(int i = 0; i < n; ++i) {
                int d = 3 * w * treasures[i].d;
                int v = treasures[i].v;
                for(int j = t; j >= d; --j) {
                    if(dp[j] < dp[j - d] + v) {
                        dp[j] = dp[j - d] + v;
                        selected[i][j] = true;
                    }
                }
            }
            output.append(dp[t]).append('\n');
            int currentT = t;
            ArrayList<Integer> treasureID = new ArrayList<>();
            for(int i = n - 1; i >= 0; --i) {
                if(selected[i][currentT]) {
                    treasureID.add(i);
                    currentT -= 3 * w * treasures[i].d;
                }
            }
            output.append(treasureID.size()).append('\n');
            for(int i = treasureID.size() - 1; i >= 0; --i)
                output.append(treasures[treasureID.get(i)].d).append(' ').append(treasures[treasureID.get(i)].v).append('\n');
            firstCase = false;
        }
	    System.out.print(output);
	}
}

class Treasure {
    int d, v;
}
